specs/interop/messaging.md

```diff
@@ -136,13 +149,121 @@ This feature could be added in a backwards compatible way by adding a new functi
+
+The expiry invariant invalidates inclusion of any executing message with
+`id.timestamp + EXPIRY_TIME < executing_block.timestamp` where:
+
+- `id` is the [`Identifier`] encoded in the executing message, matching the block attributes of the initiating message.
+- `executing_block` is the block where the executing message was included in.
+- `EXPIRY_TIME = 180 * 24 * 60 * 60 = 15552000` seconds, i.e. 180 days.
+
+## Message Graph
+
+The dependencies of messages can be modeled as a directed graph:
+
+- **vertex**: a block
+- **edge**: a dependency:
+  - parent-block (source) to block (target)
+  - message relay, from initiation (source) to execution (target)
+
+If the source of an edge is invalidated, the target is invalidated.
+
+### Invalid messages
+
+A message is said to be "invalid" when the executing message does not have a valid dependency.
+
+Dependencies are invalid when:
+
+- The dependency is unknown.
+- The dependency is known but not part of the canonical chain.
+- The dependency is known but does not match all message attributes ([`Identifier`] and payload).
+
+#### Block reorgs
+
+The [`Identifier`] used by an executing message does not cryptographically
+commit to the block it may reference.
+
+Messages may be executed before the block that initiates them is sealed.
+
+When tracking message dependencies, edges are maintained for *all* identified source blocks.
+
+Reorgs are resolved by filtering the view of the solver to only canonical blocks.
+If the source block is not canonical, the dependency is invalid.
+
+The canonical L2 block at the identified block-height is the source of truth.
+
+#### Block Recommits
+
+Blocks may be partially built, or sealed but not published, and then rebuilt.
+
+Any optimistically assumed messages have to be revalidated,
+and the original partial block can be removed from the dependency graph.
+
+### Intra-block messaging: cycles
+
+While messages cannot be initiated by future blocks,
+they can be initiated by any transactions within the same timestamp,
+as per the [Timestamp Invariant](#timestamp-invariant).
+
+This property allows messages to form cycles in the graph:
+blocks with equal timestamp, of chains in the same dependency set,
+may have dependencies on one another.
+
+### Resolving cross-chain safety
+
+To determine cross-chain safety, the graph is inspected for valid graph components that have no invalid dependencies,
+while applying the respective safety-view on the blocks in the graph.
+
+I.e. the graph must not have any inward edges towards invalid blocks within the safety-view.
+
+A safety-view is the subset of canonical blocks of all chains with the specified safety label or a higher safety label.
+Dependencies on blocks outside of the safety-view are invalid,
+but may turn valid once the safety-view changes (e.g. a reorg of unsafe blocks).
+
+By resolving in terms of graph-components, cyclic dependencies and transitive dependencies are addressed.
+
+### Horizon timestamp
+
```

Comment:
I am confused how this works in practice, like basically queue any blocks past a timestamp and then only begin to add them to the graph once the full graph with lower timestamps is solved as much as possible, only leaving a graph that has dependencies with a larger timestamp?

Reply:
Basically yes. Thanks to the timestamp invariant we can determine what is relevant to solving existing dependencies, and what can wait till later solving passes.